Objects/typeobject.c

```diff
@@ -6775,6 +6793,21 @@ type_ready_post_checks(PyTypeObject *type)
                      type->tp_name);
         return -1;
     }
+    if (type->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
+        if (type->tp_dictoffset != -1) {
+            PyErr_Format(PyExc_SystemError,
+                        "type %s has the Py_TPFLAGS_MANAGED_DICT flag "
+                        "but tp_dictoffset is set to incompatible value",
+                        type->tp_name);
+            return -1;
+        }
+    }
+    else if (type->tp_dictoffset < 0) {
+        if (type->tp_dictoffset + type->tp_basicsize <= 0) {
+            PyErr_Format(PyExc_SystemError,
+                         "type %s has a tp_dictoffset that is too small");
+        }
```

Comment:
> This is just a sanity check, right? 

For Python classes, yes. But C extensions might trigger this if they specify a too small `tp_dictoffset`

> Regardless, tp_basicsize always includes the PyObject_HEAD (or PyObject_VAR_HEAD) fields, so shouldn't that be the baseline, rather than 0?

Yes, it should. Thanks.

I'm not sure about the alignment. 
E.g.  It is theoretically possible that a class is nominally variable sized, but all instances have a size of 7.
In which case, A `tp_dictoffset` of 15 would be valid.